lambda-runtime/src/main/scala/feral/lambda/runtime/LambdaRuntime.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda
+package runtime
+
+import cats.effect.Resource
+import cats.effect.Temporal
+import cats.effect.std.Queue
+import cats.effect.syntax.resource._
+import cats.syntax.all._
+import io.circe.Decoder
+import io.circe.Encoder
+import org.http4s.client.Client
+
+import scala.util.control.NonFatal
+
+import api._
+
+object LambdaRuntime {
+
+  def apply[F[_]: Temporal: LambdaRuntimeEnv, Event: Decoder, Result: Encoder](
+      client: Client[F])(
+      handler: Resource[F, Invocation[F, Event] => F[Option[Result]]]): F[Nothing] =
+    LambdaRuntimeAPIClient(client).flatMap(client =>
+      handler.both(LambdaSettings.fromLambdaEnv.toResource).attempt.use[INothing] {
+        case Right((handler, settings)) =>
+          runloop[F, Event, Result](client, settings, handler)
+        case Left(ex) => client.reportInitError(ex) *> ex.raiseError[F, INothing]
+      })
+
+  private def runloop[F[_]: Temporal, Event: Decoder, Result: Encoder](
+      client: LambdaRuntimeAPIClient[F],
+      settings: LambdaSettings,
+      run: Invocation[F, Event] => F[Option[Result]]) = {
+    Queue.unbounded[F, LambdaRequest].flatMap { q =>
+      val producer = client
+        .nextInvocation()
+        .flatMap(q.offer)
+        .handleErrorWith {
+          case ex @ ContainerError => ex.raiseError[F, Unit]
+          case NonFatal(_) => ().pure
+          case ex => ex.raiseError
+        }
+        .foreverM[INothing]
+      val handleRequest = handleSingleRequest(client, settings, run)
+      val workers = fs2
+        .Stream
+        .fromQueueUnterminated(q)
+        .parEvalMapUnordered(1024)(handleRequest)
+        .compile
+        .drain
+      Temporal[F].both(producer, workers).map(_._1)
```

Comment:
> strange assumptions such as the use of environment variables to pass around tracing headers

Running handlers in parallel conflicts with Ray SDK assumption, but I think we could ignore X-Ray specific limitation because
- if someone is happy with official Java/JS runtime and X-Ray SDK, they can use feral-lambda-runtime-binding, both of which follow the spec to propagate X-Ray header via environment variables for each invocation
- we want to support Scala Native, which AWS does not officially support at all and thus no X-Ray SDK is available for it

Reply:
Another idea is just letting users choose qequential/concurrent execution by providing an option such as `ExecutionMode.{Parallel/Sequential}`.

Reply:
According to https://github.com/RhinoSecurityLabs/Cloud-Security-Research/blob/b7d14b7eb51b7dea6ea6a8978682bd9671e0e183/AWS/lambda_ssrf/README.md and the comment in the issue, multiple call to invocation API would return the identical results 🤔 
In addition, I couldn't find any example that runs handlers in parallel...